docs/proposals/001-proposal-merge-appset-argocd.md

```diff
@@ -0,0 +1,163 @@
+
+## Motivation
+
+Motivation is to have a tighter integration of ApplicationSet into Argo CD. Going forward, we can have first-class ApplicationSet support in Argo CD.
+
+Merging will solve this two issues
+- Currently have to Deal with two seperate docker images.After merge will have one single image
+- Circular dependency.Since Application set will not vendor argocd after merge,circular dependency will be solved.
+
+### Goals
+
+- Finalizing on how ApplicationSet to run as part of Argo CD
+- To have all tests of ApplicationSet pass when merged into Argo CD
+- Removal or refactoring of repeated code between Argo CD and ApplicationSet
+- Introduction of new tests if required 
+
+### Non-Goals
+
+ - Merging of repository of ApplicationSet so that ApplicationSet is maintained into Argo CD
+ - Pave a way to add backend support for ApplicationSet into Argo CD
+
+## Proposal
+
+Below are the possible approaches to merge ApplicationSet into Argo CD.
+
+#### Option 1 Merging ApplicationSet Controller to the same process of Argo CD Application Controller
+
+#### Merge application-set-controller into existing application-controller 
+
+Since the job of the Application Controller is to watch Applications and generate corresponding Kubernetes resources declared in Git, ApplicationSet Controller performs the same task for applications with more powerful generators. Also, it's an alternative to apps of apps pattern.
+
+Since both controllers deal with creaion/deletion of applications one should consider merging the controllers together.
+### Pros 
+-  Running ApplicationSet controller into same process as of application controller argocd, no seperate deployment is required
+-  ApplicationSet will be more tightly coupled with Application controller of argocd
+-  Can include managing of Applicationset objects into argocd api server. No need to delegate request from Argo CD to ApplicationSet Controller seperately.Delegation will involve passing of auth etc.
+-  RBAC of applications that applicatione set is managing can be inherited for ApplicatonSet RBAC
+
+### Cons
+- A lot of work. Technically difficult to merge. A lot of things in Argo CD can break.
+- Concerns wrt HA, currently workload distribution of Application controller in Argo CD is based on number of managed clusters and same can't be applied to application set. 
+
+If we consider scaling beyond number of managed clusters, we are considering proposal of sharding based on number of applications in argocd and each replica to manage shards with some kind of election. Concerns wrt HA can be overcome and same can be applied to ApplicationSet CRD  but need to consider it implementing along with merging which can be challenging.
+
+
+#### Option 2 Run ApplicationSet as a seperate process/service
```

Comment:
I think we agreed that this option is preferable. It is significantly less work and users have the option to "disable the" feature if they don't use it. @jessesuen , @jannfis , @jgwest please comment if you think otherwise.

If everyone agrees please let explicitly mention that this option is preferable.

Reply:
Agreed this is definitely preferable as our first step. At a later point, for further code-level integration is useful, that can be considered then.